OAuth: more debug logs, configurable browser

woodruffw opened the pull request:
~~WIP.~~

~~This adds more debug logs and adds the `SIGSTORE_OAUTH_BROWSER` environment variable, which can be used to override the `webbrowser`'s default browser selection (e.g. `SIGSTORE_OAUTH_BROWSER=firefox`).~~

Removed the `SIGSTORE_OAUTH_BROWSER` variable, since it's essentially a duplicate of what `webbrowser` already does via the `BROWSER` variable. Instead, this PR adds `SIGSTORE_OAUTH_FORCE_OOB`, which forcefully disables the in-band OAuth flow if set.

See #96.

kushaldas commented:
I am seeing: 

I replaced the real values for `{AUTH_CODE_VALUE}&state={MYSTATE_VALUE}`


```
DEBUG:sigstore._internal.oidc.oauth:GET: /auth/callback?code={AUTH_CODE_VALUE}&state={MYSTATE_VALUE}
DEBUG:sigstore._internal.oidc.oauth:GET: /favicon.ico
DEBUG:sigstore._internal.oidc.oauth:GET: /favicon.ico
DEBUG:sigstore._internal.oidc.oauth:GET: /favicon.ico
DEBUG:sigstore._internal.oidc.oauth:GET: /favicon.ico
DEBUG:sigstore._internal.oidc.oauth:GET: /favicon.ico
DEBUG:sigstore._internal.oidc.oauth:GET: /favicon.ico
DEBUG:sigstore._internal.oidc.oauth:GET: /favicon.ico
DEBUG:sigstore._internal.oidc.oauth:GET: /favicon.ico
DEBUG:sigstore._internal.oidc.oauth:GET: /favicon.ico
DEBUG:sigstore._internal.oidc.oauth:GET: /favicon.ico
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): oauth2.sigstore.dev:443
DEBUG:urllib3.connectionpool:https://oauth2.sigstore.dev:443 "POST /auth/token HTTP/1.1" 400 70
Traceback (most recent call last):
  File "/home/kdas/code/sigstore-python/sigstore/_internal/oidc/oauth.py", line 211, in get_identity_token
    resp.raise_for_status()
  File "/home/kdas/code/sigstore-python/.venv/lib64/python3.10/site-packages/requests/models.py", line 1022, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 400 Client Error: Bad Request for url: https://oauth2.sigstore.dev/auth/token

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/lib64/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib64/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/home/kdas/code/sigstore-python/sigstore/__main__.py", line 22, in <module>
    main()
  File "/home/kdas/code/sigstore-python/sigstore/_cli.py", line 254, in main
    _sign(args)
  File "/home/kdas/code/sigstore-python/sigstore/_cli.py", line 333, in _sign
    args.identity_token = get_identity_token(
  File "/home/kdas/code/sigstore-python/sigstore/_internal/oidc/oauth.py", line 213, in get_identity_token
    raise IdentityError from http_error
sigstore._internal.oidc.IdentityError

```

woodruffw commented:
Thanks. That series of favicon requests looks suspiciously like the race condition we thought we fixed earlier. 

Sent from mobile. Please excuse my brevity.

> On Jun 21, 2022, at 12:39 PM, Kushal Das ***@***.***> wrote:
> 
> ﻿
> I am seeing:
> 
> I replaced the real values for {AUTH_CODE_VALUE}&state={MYSTATE_VALUE}
> 
> DEBUG:sigstore._internal.oidc.oauth:GET: /auth/callback?code={AUTH_CODE_VALUE}&state={MYSTATE_VALUE}
> DEBUG:sigstore._internal.oidc.oauth:GET: /favicon.ico
> DEBUG:sigstore._internal.oidc.oauth:GET: /favicon.ico
> DEBUG:sigstore._internal.oidc.oauth:GET: /favicon.ico
> DEBUG:sigstore._internal.oidc.oauth:GET: /favicon.ico
> DEBUG:sigstore._internal.oidc.oauth:GET: /favicon.ico
> DEBUG:sigstore._internal.oidc.oauth:GET: /favicon.ico
> DEBUG:sigstore._internal.oidc.oauth:GET: /favicon.ico
> DEBUG:sigstore._internal.oidc.oauth:GET: /favicon.ico
> DEBUG:sigstore._internal.oidc.oauth:GET: /favicon.ico
> DEBUG:sigstore._internal.oidc.oauth:GET: /favicon.ico
> DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): oauth2.sigstore.dev:443
> DEBUG:urllib3.connectionpool:https://oauth2.sigstore.dev:443 "POST /auth/token HTTP/1.1" 400 70
> Traceback (most recent call last):
>   File "/home/kdas/code/sigstore-python/sigstore/_internal/oidc/oauth.py", line 211, in get_identity_token
>     resp.raise_for_status()
>   File "/home/kdas/code/sigstore-python/.venv/lib64/python3.10/site-packages/requests/models.py", line 1022, in raise_for_status
>     raise HTTPError(http_error_msg, response=self)
> requests.exceptions.HTTPError: 400 Client Error: Bad Request for url: https://oauth2.sigstore.dev/auth/token
> 
> The above exception was the direct cause of the following exception:
> 
> Traceback (most recent call last):
>   File "/usr/lib64/python3.10/runpy.py", line 196, in _run_module_as_main
>     return _run_code(code, main_globals, None,
>   File "/usr/lib64/python3.10/runpy.py", line 86, in _run_code
>     exec(code, run_globals)
>   File "/home/kdas/code/sigstore-python/sigstore/__main__.py", line 22, in <module>
>     main()
>   File "/home/kdas/code/sigstore-python/sigstore/_cli.py", line 254, in main
>     _sign(args)
>   File "/home/kdas/code/sigstore-python/sigstore/_cli.py", line 333, in _sign
>     args.identity_token = get_identity_token(
>   File "/home/kdas/code/sigstore-python/sigstore/_internal/oidc/oauth.py", line 213, in get_identity_token
>     raise IdentityError from http_error
> sigstore._internal.oidc.IdentityError
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were assigned.

woodruffw commented:
@kushaldas could you try once more with the changes in [91b22a0](https://github.com/sigstore/sigstore-python/pull/136/commits/91b22a0bbd88261686c672f900c87c29953762f7)? That should prevent any races if the browser attempts to spam us with favicon requests.

woodruffw commented:
This should be good to go. It "fixes" the idempotency issue by doing one less call to `server.auth_request()`. That's not ideal in the long term, however, since the API is still pretty easy to misuse. We should think about making it more misuse-resistant.

di commented on sigstore/_internal/oidc/oauth.py:
```diff
@@ -48,16 +51,21 @@ def log_message(self, _format: str, *_args: Any) -> None:
         pass
 
     def do_GET(self) -> None:
+        logger.debug(f"GET: {self.path}")
```
We should include headers and other details about the request here as well (query params?)

woodruffw replied:
Yeah, good idea. In the "normal" case there should be very little here (since we only handle the initial kickoff that redirects to sigstore + the response), but it'll help debug pathological cases.

woodruffw replied:
(The query params will already be present, since they're embedded in `self.path`)

woodruffw replied:
[96b93bd](https://github.com/sigstore/sigstore-python/pull/136/commits/96b93bda456078bc431da8927c972b2e383b5f85) adds the headers to the log.

(Tangent: we should think about how/if we want to warn users about potentially sensitive materials in their logs.)
